### PR TITLE
Add `token_program` constraint to token, mint, and associated token accounts

### PR DESCRIPTION
**Pull request by elliotkennedy**
### Overview

It is currently only possible to cpi a single `token_program`.

As discussed in https://github.com/coral-xyz/anchor/pull/2363 & https://github.com/coral-xyz/anchor/pull/2386 - it should be possible to override the required `token_program` account field when intializing a `Mint`, `TokenAccount` or `AssociatedTokenAccount`.

These new constraints allow this, and also act as `owner` constraints when not used with an `init`.

This is useful when an instruction needs to support some combination of spl-token and spl-token-2022 mints.

### API changes

- Added a new `token_program` attribute to `TokenAccount`, `Mint` and `AssociatedTokenAccount` types - this allows you to specify the spl-token or spl-token-2022 `token_program` field which should be used in constraints and CPI calls - rather than enforcing a single account field called `token_program`.

Example skeleton endpoint supporting both spl-token and spl-token-2022 mints -

```rs
#[derive(Accounts)]
pub struct InitializePool<'info> {

    /// Token A mint
    #[account(
        mint::token_program = token_a_token_program,
    )]
    pub token_a_mint: Box<InterfaceAccount<'info, Mint>>,
    /// Token B mint
    #[account(
        mint::token_program = token_b_token_program,
    )]
    pub token_b_mint: Box<InterfaceAccount<'info, Mint>>,

    #[account(init,
        seeds = [b"pvault_a".as_ref(), pool.key().as_ref(), token_a_mint.key().as_ref()],
        bump,
        payer = admin_authority,
        token::mint = token_a_mint,
        token::authority = pool_authority,
        token::token_program = token_a_token_program,
    )]
    pub token_a_vault: Box<InterfaceAccount<'info, TokenAccount>>,

    #[account(init,
        seeds = [b"pvault_b".as_ref(), pool.key().as_ref(), token_b_mint.key().as_ref()],
        bump,
        payer = admin_authority,
        token::mint = token_b_mint,
        token::authority = pool_authority,
        token::token_program = token_b_token_program,
    )]
    pub token_b_vault: Box<InterfaceAccount<'info, TokenAccount>>,

    pub system_program: Program<'info, System>,
    /// The token program for the token A mint
    pub token_a_token_program: Interface<'info, TokenInterface>,
    /// The token program for the token B mint
    pub token_b_token_program: Interface<'info, TokenInterface>,
}
```

### Discussion

- In most cases, the owner constraint would be sufficient here, however because of the spl-token cpi call integration in anchor, and the existing requirement to specify a `token_program` field; a new, specialized, attribute seems correct for this.

### Breaking changes

None, where the attribute is not used, the existing `token_program` account field will be used.

### CODE REVIEW

File: lang/syn/src/lib.rs [hunk 1]
```diff
@@ -662,11 +662,14 @@ pub enum ConstraintToken {
     Address(Context<ConstraintAddress>),
     TokenMint(Context<ConstraintTokenMint>),
     TokenAuthority(Context<ConstraintTokenAuthority>),
+    TokenTokenProgram(Context<ConstraintTokenProgram>),
     AssociatedTokenMint(Context<ConstraintTokenMint>),
     AssociatedTokenAuthority(Context<ConstraintTokenAuthority>),
+    AssociatedTokenProgram(Context<ConstraintTokenProgram>),
```
Comment (acheroncrypto) [inline lang/syn/src/lib.rs:668]: Same here, consistency with `TokenTokenProgram` above and make it more clear.

```suggestion
    AssociatedTokenTokenProgram(Context<ConstraintTokenProgram>),
```

File: lang/syn/src/parser/accounts/constraints.rs [hunk 2]
```diff
@@ -332,11 +350,14 @@ pub struct ConstraintGroupBuilder<'ty> {
     pub address: Option<Context<ConstraintAddress>>,
     pub token_mint: Option<Context<ConstraintTokenMint>>,
     pub token_authority: Option<Context<ConstraintTokenAuthority>>,
+    pub token_token_program: Option<Context<ConstraintTokenProgram>>,
     pub associated_token_mint: Option<Context<ConstraintTokenMint>>,
     pub associated_token_authority: Option<Context<ConstraintTokenAuthority>>,
+    pub associated_token_program: Option<Context<ConstraintTokenProgram>>,
```
Comment (acheroncrypto) [inline lang/syn/src/parser/accounts/constraints.rs:356]: Same here, consistency with `token_token_program` above and to make it more clear.

```suggestion
    pub associated_token_token_program: Option<Context<ConstraintTokenProgram>>,
```

File: lang/syn/src/parser/accounts/constraints.rs [hunk 3]
```diff
@@ -988,6 +1057,31 @@ impl<'ty> ConstraintGroupBuilder<'ty> {
...
+            return Err(ParseError::new(
+                c.span(),
+                "token token_program already provided",
+            ));
+        }
+        self.token_token_program.replace(c);
+        Ok(())
+    }
+
+    fn add_associated_token_program(
```
Comment (acheroncrypto) [inline lang/syn/src/parser/accounts/constraints.rs:1071]: Same here, consistency with `add_token_token_program` above and to make it more clear.

```suggestion
    fn add_associated_token_token_program(
```

File: lang/syn/src/lib.rs [hunk 4]
```diff
@@ -844,6 +855,11 @@ pub struct ConstraintMintDecimals {
     pub decimals: Expr,
 }
 
+#[derive(Debug, Clone)]
+pub struct ConstraintMintTokenProgram {
```
Comment (acheroncrypto) [inline lang/syn/src/lib.rs:859]: Is there any advantage of defining a new constraint here and not use `ConstraintTokenProgram`?

**Review by acheroncrypto — COMMENTED** [review]
Looking great, please see the comments.

File: lang/src/error.rs [hunk 5]
```diff
@@ -110,6 +110,17 @@ pub enum ErrorCode {
...
+    ///
+    /// 2021 - A token account token program constraint was violated
+    #[msg("A token account token program constraint was violated")]
+    ConstraintTokenTokenProgram,
+    /// 2022 - A mint token program constraint was violated
+    #[msg("A mint token program constraint was violated")]
+    ConstraintMintTokenProgram,
+    /// 2023 - A mint token program constraint was violated
+    #[msg("An associated token account token program constraint was violated")]
+    ConstraintAssociatedTokenProgram,
```
[the remark this answers is not in the document]
Comment (elliotkennedy) [reply lang/src/error.rs:123]: Yes I think that is clearer too

File: lang/syn/src/parser/accounts/constraints.rs (hunk 2 above)
Comment (elliotkennedy) [reply lang/syn/src/parser/accounts/constraints.rs:356]: Sounds good

File: lang/syn/src/parser/accounts/constraints.rs (hunk 3 above)
Comment (elliotkennedy) [reply lang/syn/src/parser/accounts/constraints.rs:1071]: Yep, good spot

File: lang/syn/src/lib.rs (hunk 1 above)
Comment (elliotkennedy) [reply lang/syn/src/lib.rs:668]: Sounds good

File: lang/syn/src/lib.rs (hunk 4 above)
Comment (elliotkennedy) [reply lang/syn/src/lib.rs:859]: I thought I was following a convention - but on second look I think reusing the same struct is fine.

**Comment by elliotkennedy** [comment]
> Looking great, please see the comments.

Thanks so much for your review @acheroncrypto - on addressing the comments I realised that the initialized `Mint` constraint was not implemented - I have added this in the latest commit. I am currently writing some tests around these constraints, previously I was not aware of the misc test suite and it looks ideal. Will ping when ready for another look.

**Comment by elliotkennedy** [comment]
Hey @acheroncrypto, I added a lot of test cases in the last commit - they are summarised in the table below:

|       | `init` successful initialize | Constraint (no init) pass case | Constraint (no init) error case | `init-if-needed` successful initialize | `init-if-needed` constraint (no init) pass case | `init-if-needed` constraint (no init) error case |
|-------|------------------------------|--------------------------------|---------------------------------|----------------------------------------|-------------------------------------------------|--------------------------------------------------|
| Mint  | :heavy_check_mark:           | :heavy_check_mark:             | :heavy_check_mark:              | :heavy_check_mark:                     | :heavy_check_mark:                              | :heavy_check_mark:                               |
| Token | :heavy_check_mark:           | :heavy_check_mark:             | :heavy_check_mark:              | :heavy_check_mark:                     | :heavy_check_mark:                              | :heavy_check_mark:                               |
| ATA   | :heavy_check_mark:           | :heavy_check_mark:             | :heavy_check_mark:              | :heavy_check_mark:                     | :heavy_check_mark:                              | :heavy_check_mark:                               |

I fixed a couple of bugs in the process, mostly around missing `init-if-needed` constraints.

I did not fix https://github.com/coral-xyz/anchor/pull/2447 and there are no test cases around ATA initialization with `token-2022`; however the token account and mint `token-2022` test cases are covered in the new `token-wrapper` test suite.

I think this is now good to ship - please take another look when you can - thanks!

**Comment by acheroncrypto** [comment]
> In most cases, the owner constraint would be sufficient here, however because of the spl-token cpi call integration in anchor, and the existing requirement to specify a `token_program` field; a new, specialized, attribute seems correct for this.

As you've stated, there are some cases that wouldn't work with just an owner constraint but I have to say this does feel a bit of an anti-pattern that we are specifying the program when using `InterfaceAccount` since the whole point of `InterfaceAccount` is to deserialize into the given type with multiple owners safely. Not a problem with this PR but I'd like to know your opinion on this too.

> on addressing the comments I realised that the initialized `Mint` constraint was not implemented

Not specifically speaking of this PR but there is quite a bit of duplicate code on these checks which would make us much more likely to miss some of the implementations.

> I fixed a couple of bugs in the process, mostly around missing `init-if-needed` constraints.

Thank you for fixing those and also for writing tests that cover many cases.

**Comment by elliotkennedy** [comment]
> > > In most cases, the owner constraint would be sufficient here, however because of the spl-token cpi call integration in anchor, and the existing requirement to specify a `token_program` field; a new, specialized, attribute seems correct for this.
> >
> > As you've stated, there are some cases that wouldn't work with just an owner constraint but I have to say this does feel a bit of an anti-pattern that we are specifying the program when using `InterfaceAccount` since the whole point of `InterfaceAccount` is to deserialize into the given type with multiple owners safely. Not a problem with this PR but I'd like to know your opinion on this too.
> >

I take your point ;) - however I think of this as more of a reference from an `InterfaceAccount` -> `Interface`; or perhaps a reference to a runtime label in cases where the implementation is ambiguous and can't be inferred safely.

My opinion is that in general there should be a preference for inferring the interface implementation but any time you want to mix implementations of the same interface you will need some some sort of type reference. Particularly with Solana requiring program state up-front.

I think there is a wider question about how interfaces should work in Solana and that probably needs to be addressed in the runtime. I believe it is on the roadmap for v2 but I don't have any idea how it should look. However it is a distinct problem to the current anchor notion of interfaces.

Another thought is perhaps an `InterfaceAccount` specific new `impl/implementation` constraint would be preferable to reusing `owner` but I don't think there is a requirement for it yet.

Another approach would be to inverse the reference/constraint - `Interface` -> `InterfaceAccount` but I don't think this would be as intuitive.

I think removing the `token_program` convention would be less clean for anchor code in general. Perhaps you could generify the `token_program` field conventional approach by adding a `program_field()` method to the `Interface` accounts. This would return the string field name which needs to be added to the `Accounts` struct. You would also need a constraint to activate it as the `InterfaceAccount` does not always require the `Interface`.

There's another question about new token 2022 features - how should they be implemented? Do we continue adding constraints to initialise Mints with new flags? 

> > > on addressing the comments I realised that the initialized `Mint` constraint was not implemented
> >
> > Not specifically speaking of this PR but there is quite a bit of duplicate code on these checks which would make us much more likely to miss some of the implementations.
> >

I agree :) I think there is definitely scope for reuse around `init-if-needed` and `init` - it is too easy to miss atm as they are complete reimplementations of the same thing.

> > > I fixed a couple of bugs in the process, mostly around missing `init-if-needed` constraints.
> >
> > Thank you for fixing those and also for writing tests that cover many cases.
> >

Pleasure + very, very relieved to look in the misc package :)

**Review by acheroncrypto — APPROVED** [review]
Thanks again for such a great PR! It's well written, no breaking changes, fixes bugs and adds many test cases. 👏

**Comment by Dodecahedr0x** [comment]
I'm trying to use this in my project but I get a `Signature verification failed` error coming from the token account.

I create it like so:
```
#[account(
    init,
    payer = payer,
    token::mint = mint,
    token::authority = creator,
    token::token_program = token_program
)]
pub token_account: InterfaceAccount<'info, TokenAccount>,
```

And removing it makes my instruction work. I suppose I need to sign the IX using the token account keys but I don't have them since I want an ATA PDA.

Is this expected behavior? Is there a way around it?

**Comment by elliotkennedy** [comment]
> I'm trying to use this in my project but I get a `Signature verification failed` error coming from the token account.
> 
> I create it like so:
> 
> ```
> #[account(
>     init,
>     payer = payer,
>     token::mint = mint,
>     token::authority = creator,
>     token::token_program = token_program
> )]
> pub token_account: InterfaceAccount<'info, TokenAccount>,
> ```
> 
> And removing it makes my instruction work. I suppose I need to sign the IX using the token account keys but I don't have them since I want an ATA PDA.
> 
> Is this expected behavior? Is there a way around it?

This is expected. You need to use the `AssociatedToken` type to get anchor to invoke the correct CPIs iirc.

Try -

```rs
#[account(
    init,
    payer = payer,
    token::mint = mint,
    token::authority = creator,
    token::token_program = token_program
)]
pub token_account: InterfaceAccount<'info, AssociatedToken>,
```

**Comment by Dodecahedr0x** [comment]
Thanks for your answer!

Using `anchor_spl::associated_token::AssociatedToken` gets me this error:
```
error[E0277]: the trait bound `AssociatedToken: anchor_lang::AccountSerialize` is not satisfied
   --> programs/nft-standard/src/instructions/create_metadata.rs:84:24
    |
84  |     pub token_account: InterfaceAccount<'info, AssociatedToken>,
    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `anchor_lang::AccountSerialize` is not implemented for `AssociatedToken`
    |
    = help: the following other types implement trait `anchor_lang::AccountSerialize`:
              AuthoritiesGroup
              UpgradeableLoaderState
              __idl::IdlAccount
              anchor_lang::ProgramData
              anchor_lang::idl::IdlAccount
              anchor_spl::token::Mint
              anchor_spl::token::TokenAccount
              anchor_spl::token_interface::Mint
            and 2 others
note: required by a bound in `anchor_lang::prelude::InterfaceAccount`
   --> /home/dode/.cargo/registry/src/index.crates.io-6f17d22bba15001f/anchor-lang-0.28.0/src/accounts/interface_account.rs:160:39
    |
160 | pub struct InterfaceAccount<'info, T: AccountSerialize + AccountDeserialize + Clone> {
    |                                       ^^^^^^^^^^^^^^^^ required by this bound in `InterfaceAccount`
```

And another saying `field `owner` of `InterfaceAccount` is private`

**Comment by elliotkennedy** [comment]
> Thanks for your answer!
> 
> Using `anchor_spl::associated_token::AssociatedToken` gets me this error:
> 
> ```
> error[E0277]: the trait bound `AssociatedToken: anchor_lang::AccountSerialize` is not satisfied
>    --> programs/nft-standard/src/instructions/create_metadata.rs:84:24
>     |
> 84  |     pub token_account: InterfaceAccount<'info, AssociatedToken>,
>     |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `anchor_lang::AccountSerialize` is not implemented for `AssociatedToken`
>     |
>     = help: the following other types implement trait `anchor_lang::AccountSerialize`:
>               AuthoritiesGroup
>               UpgradeableLoaderState
>               __idl::IdlAccount
>               anchor_lang::ProgramData
>               anchor_lang::idl::IdlAccount
>               anchor_spl::token::Mint
>               anchor_spl::token::TokenAccount
>               anchor_spl::token_interface::Mint
>             and 2 others
> note: required by a bound in `anchor_lang::prelude::InterfaceAccount`
>    --> /home/dode/.cargo/registry/src/index.crates.io-6f17d22bba15001f/anchor-lang-0.28.0/src/accounts/interface_account.rs:160:39
>     |
> 160 | pub struct InterfaceAccount<'info, T: AccountSerialize + AccountDeserialize + Clone> {
>     |                                       ^^^^^^^^^^^^^^^^ required by this bound in `InterfaceAccount`
> ```
> 
> And another saying `field 'owner' of 'InterfaceAccount' is private`

Apologies I misremembered the syntax there - 

Try this:

```rs
#[account(
    init,
    payer = payer,
    associated_token::mint = mint,
    associated_token::authority = creator,
    associated_token::token_program = token_program
)]
pub token_account: InterfaceAccount<'info, TokenAccount>,

pub associated_token_program: Program<'info, AssociatedToken>,
```

**Comment by Dodecahedr0x** [comment]
It worked! I was missing the associated program but the error got me confused..

Thanks a lot, I will point to this thread on StacckExchange

**Comment by CeTesDev** [comment]
Hi, I have a similar issue on my program.
I am creating the program to transfer token(normal and extension both). 

```
#[derive(Accounts)]
pub struct DepositToken<'info> {
    #[account(mut)]
    /// Payer will pay Wormhole fee to post a message.
    pub payer: Signer<'info>,

    #[account(
        mut,
        token::mint = mint,
        token::authority = payer,
        token::token_program = token_program
    )]
    pub from_token_account: InterfaceAccount<'info, TokenAccount>,

    #[account(
        // mut,
        init_if_needed,
        payer = payer,
        token::mint = mint,
        token::authority = dest_account,
        token::token_program = token_program
    )]
    pub to_token_account: InterfaceAccount<'info, TokenAccount>,

    /// CHECK: This is not dangerous because we will compare it with dest_addr in code
    #[account(mut)]
    pub dest_account: AccountInfo<'info>,

    /// System program.
    pub system_program: Program<'info, System>,

    /// Token program.
    pub mint: InterfaceAccount<'info, Mint>,

    /// Token program.
    pub token_program: Interface<'info, TokenInterface>,

    /// CHECK: This is not dangerous because we don't read or write from this account~
    pub associated_token_program: Program<'info, AssociatedToken>,

    /// Clock sysvar.
    pub clock: Sysvar<'info, Clock>,

    /// Rent sysvar.
    pub rent: Sysvar<'info, Rent>,
}
```

I have an error "Error: Signature verification failed".

Could you help me?
Regards

**Pull request merged**